### PR TITLE
Removed assumptions about ancestry from fork tree prune method

### PR DESCRIPTION
**Pull request by davxy**
Fork tree prune method was **assuming** that the user's `predicate` closure makes the `find_node_index_where` to **NOT** return the tree node immediately preceding the `hash` parameter (according to `is_descendent_of` closure).

Up to date the only users of this structure are GRANDPA and BABE to keep a tree of epoch changes announcements AND according to their `predicate`, the assumption is satisfied.

But this is a **time-bomb**.

If at some point the struct is used in a context where the predicate doesn't satisfy the assumption, the tree would break and we loose part of the tree due to how `prune` logic is implemented.
(indeed this assumption was the cause of https://github.com/paritytech/substrate/pull/12758 bug)

This PR removes the assumption and makes the tree more resilient and independent for any use case.

---

- [x] Burn-in test

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2084698

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2084697

File: utils/fork-tree/src/lib.rs [hunk 1]
```diff
@@ -351,14 +354,16 @@ where
 		})
 	}
 
-	/// Prune the tree, removing all non-canonical nodes. We find the node in the
-	/// tree that is the deepest ancestor of the given hash and that passes the
-	/// given predicate. If such a node exists, we re-root the tree to this
-	/// node. Otherwise the tree remains unchanged. The given function
-	/// `is_descendent_of` should return `true` if the second hash (target) is a
-	/// descendent of the first hash (base).
+	/// Prune the tree, removing all non-canonical nodes.
```
Comment (davxy) [inline utils/fork-tree/src/lib.rs:357]: The PR mostly involves the `ForkTree::prune` method.

The rest of the PR are mostly tests

**Comment by davxy** [comment]
A burn-in has been performed. Both full sync and warp went good

**Comment by bkchr** [comment]
> A burn-in has been performed. Both full sync and warp went good

If it worked, why did you closed the pr? :D

**Comment by davxy** [comment]
Jeez that was not intentional :-D

File: utils/fork-tree/src/lib.rs [hunk 2]
```diff
@@ -32,7 +32,7 @@ pub enum Error<E> {
 	UnfinalizedAncestor,
 	/// Imported or finalized node that is an ancestor of previously finalized node.
 	Revert,
-	/// Error throw by client when checking for node ancestry.
+	/// Error throw by user when checking for node ancestry.
```
Comment (melekes) [inline utils/fork-tree/src/lib.rs:35]:
```suggestion
	/// Error thrown by user when checking for node ancestry.
```

**Review by melekes — APPROVED** [review]
👍 (code-wise; don't know enough to reason about business logic)

**Comment by davxy** [comment]
Not stale

File: utils/fork-tree/src/lib.rs [hunk 3]
```diff
@@ -377,36 +382,57 @@ where
...
-				is_first = false;
-			} else {
-				removed.push(child);
+		// Curr now points to our direct ancestor, if necessary remove any node that is
+		// not a descendant of `hash`.
+		if curr.children.len() > 0 {
+			let children = std::mem::take(&mut curr.children);
+			for child in children {
+				if child.number == *number && child.hash == *hash ||
+					*number < child.number && is_descendent_of(hash, &child.hash)?
```
Comment (skunert) [inline utils/fork-tree/src/lib.rs:427]: nit: Maybe one comment here that this condition is inverted and we are now searching for the prune "target" or a node even deeper than the "target".

**Review by skunert — APPROVED** [review]
Logic looks good! Only super nits.

**Pull request closed without merging**